### PR TITLE
Teacher Tool: Import and Export Rubric

### PR DESCRIPTION
**Pull request by thsparks**
This change enables exporting rubrics to a file and importing them again. I've tucked the options behind a new "Action Menu", which is just a meatball menu on the right side of the toolbar. I didn't spend a ton of time styling that, just wanted to get something fairly quick working.

I'm not in love with the way the import modal looks. Definitely open to recommendations on how to make that look nicer, but this was as good as I could get it without feedback :)

Screenshots:
image

image

image

image

Upload Target: https://makecode.microbit.org/app/ef53963093d3ce1b2679fe56fa95bc3497df8e51-cefc5556d6--eval

### CODE REVIEW

File: teachertool/src/transforms/getRubricFromFileAsync.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+export async function getRubricFromFileAsync(file: File, allowPartial: boolean): Promise<Rubric | undefined> {
+    let rubric = await loadRubricFromFileAsync(file);
+
+    if (rubric) {
+        logDebug(`Loading rubric '${rubric.name}' from file...`);
+
+        const rubricVerificationResult = verifyRubricIntegrity(rubric);
+
+        if (!rubricVerificationResult.valid) {
+            postNotification(makeNotification(lf("Rubric contains invalid criteria."), 2000));
```
Comment (thsparks) [inline teachertool/src/transforms/getRubricFromFileAsync.ts:17]: Probably remove the notifications in this file. They aren't very visible and modal shows an error anyway.

File: teachertool/src/components/ImportRubricModal.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            setSelectedRubric(undefined);
+            setErrorMessage(undefined);
+        }
+    }, [selectedFile]);
+
+    function closeModal() {
+        setSelectedFile(undefined);
+        setErrorMessage(undefined);
+        setSelectedRubric(undefined);
+        hideModal("import-rubric");
```
Comment (eanders-ms) [inline teachertool/src/components/ImportRubricModal.tsx:44]: I wonder if hideModal shouldn't take a parameter. Just have it hide the current modal.

File: teachertool/src/components/ImportRubricModal.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    return teacherTool.modal === "import-rubric" ? (
+        <Modal title={lf("Select rubric to import")} actions={actions} onClose={closeModal}>
+            <div className={css["import-rubric"]}>
+                <NoticeLabel severity="warning">
+                    {lf("Warning! Your current rubric will be overwritten by the imported rubric.")}
+                </NoticeLabel>
+                {errorMessage && <NoticeLabel severity="error">{errorMessage}</NoticeLabel>}
+                {selectedRubric && (
+                    <div className={css["rubric-preview-container"]}>
```
Comment (eanders-ms) [inline teachertool/src/components/ImportRubricModal.tsx:85]: This css name is a bit overly descriptive given that it is already scoped to this module. No need to ensure it is globally unique. You could go with simply `container`.

File: teachertool/src/services/fileSystemService.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+// Serializes the active rubric and writes it to a file.
+// Returns true if the file was written successfully, false otherwise.
+export function writeRubricToFile(rubric: Rubric): boolean {
+    if (!rubric.name || rubric.criteria.length === 0) {
+        return false;
+    }
+
+    const fileName = sanitizeFileName(rubric.name) + '.json';
```
Comment (eanders-ms) [inline teachertool/src/services/fileSystemService.ts:22]: Can `rubric.name` be empty?

**Comment by eanders-ms** [comment]
It's great to get this functionality in, but there's a lot of UI here that needs design. I question the need for the preview.

File: teachertool/src/services/fileSystemService.ts (hunk 4 above)
Comment (eanders-ms) [reply teachertool/src/services/fileSystemService.ts:22]: Oh, I see the conditional above. Let's not make a name strictly required. Have it assign a default name to the file if empty: "unnamed-rubric.json".

File: teachertool/src/transforms/showModal.ts [hunk 5]
```diff
@@ -1,7 +1,8 @@
 import { stateAndDispatch } from "../state";
 import * as Actions from "../state/actions";
+import { ModalType } from "../types";
 
-export function showCatalogModal() {
+export function showModal(modal: ModalType) {
```
[the remark this answers is not in the document]
Comment (thsparks) [reply teachertool/src/transforms/showModal.ts:5]: `hideModal` also accepts a `modalType` and just uses it to verify you're hiding the modal you think you are. It's probably unnecessary validation, but that's how it works currently. Could remove the check if desired. I don't think it's redundant though, I peronally feel that having a `hide` with the corresponding `show` is nicer than using `showModal(undefined)`

[the remark this answers is not in the document]
Comment (eanders-ms) [reply teachertool/src/transforms/showModal.ts:5]: I didn't see a corresponding change to hideModal to check that the expected modal is "import-rubric". Maybe I missed it.

[the remark this answers is not in the document]
Comment (thsparks) [reply teachertool/src/transforms/showModal.ts:5]: It was already working that way. This change was bringing `showModal` into alignment with that functionality.

[the remark this answers is not in the document]
Comment (thsparks) [reply teachertool/src/transforms/showModal.ts:5]: https://github.com/microsoft/pxt/blob/thsparks/teacher_tool/import_export_rubric/teachertool/src/transforms/hideModal.ts#L9

**Review by srietkerk — COMMENTED** [review]
Just a few questions

**Comment by srietkerk** [comment]
> It's great to get this functionality in, but there's a lot of UI here that needs design. I question the need for the preview.

I really love the preview. I think, like Thomas said, it gives peace of mind to the user. 

I'm also thinking that if, in the future, we want to supply pre-built rubrics, we might be able to put them here. There might be a better place for it, but it was something I thought of as I was playing around with this. In a pre-built case, you'd want to show the user a preview before importing.

File: teachertool/src/components/DebugInput.tsx [hunk 6]
```diff
@@ -1,11 +1,10 @@
...
 
-import { useState } from "react";
+import { useContext } from "react";
 import { Button } from "react-common/components/controls/Button";
-import { Input } from "react-common/components/controls/Input";
-import { Textarea } from "react-common/components/controls/Textarea";
-import { loadProjectMetadataAsync } from "../transforms/loadProjectMetadataAsync";
 import { runEvaluateAsync } from "../transforms/runEvaluateAsync";
+import { writeRubricToFile } from "../services/fileSystemService";
+import { AppStateContext } from "../state/appStateContext";
```
[the remark this answers is not in the document]
Comment (thsparks) [reply teachertool/src/components/DebugInput.tsx:7]: Nope just leftovers from before I moved import/export. I'll remove them.

File: teachertool/src/components/NoticeLabel.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        case "neutral":
+        default:
+            // no icon
+            break;
+    }
+
+    return (
+        <div className={css["notice-label-background"]}>
+            <div className={classList(css["notice-label-container"], css[`${props.severity}-notice-label-container`])}>
+                {props.severity !== "info" && <i className={classList(iconClass, css["icon"])} />}
```
[the remark this answers is not in the document]
Comment (thsparks) [reply teachertool/src/components/NoticeLabel.tsx:32]: Ah good catch, this should be neutral not info :)

File: teachertool/src/state/helpers.ts [hunk 8]
```diff
@@ -1,7 +1,39 @@
...
+import { ErrorCode } from "../types/errorCode";
+import { Rubric } from "../types/rubric";
 import { stateAndDispatch } from "./appStateContext";
 
 export function getCatalogCriteriaWithId(id: string): CatalogCriteria | undefined {
     const { state } = stateAndDispatch();
     return state.catalog?.find(c => c.id === id);
 }
+
+export function verifyCriteriaInstanceIntegrity(instance: CriteriaInstance) {
```
[the remark this answers is not in the document]
Comment (thsparks) [reply teachertool/src/state/helpers.ts:12]: The function accepts a criteria instance and verifies that that instance's information (namely the catalog id and parameters) map to actual values in the catalog, so it's a bit of both, but from a caller's perspective, it's a scenario of "I have a criteria instance and I want to verify it's correctly defined."

I'm open to alternative name suggestions, but I did want to be careful to avoid possible confusion around verifying the instance's definition (which is what we're doing here) vs actually running validation of the code based on the criteria (hence weird "integrity" names).

[the remark this answers is not in the document]
Comment (srietkerk) [reply teachertool/src/state/helpers.ts:12]: Oh, thanks for the clarification. The confusion I had was regarding testing the rubric integrity of an uploaded rubric. I was thinking that the rubric file wouldn't have instanced criteria, but that wouldn't make sense. Thanks!

**Review by eanders-ms — APPROVED** [review]
LGTM. Will want to refine these dialogs, but great to get this in.

File: teachertool/src/services/fileSystemService.ts (hunk 4 above)
Comment (eanders-ms) [reply teachertool/src/services/fileSystemService.ts]: Another approach: give the rubric a default name from the beginning, and not let it become null/empty. Either way, I think we should provide a filename if necessary to allow the export to succeed.

**Comment by eanders-ms** [comment]
> > It's great to get this functionality in, but there's a lot of UI here that needs design. I question the need for the preview.
> 
> I really love the preview. I think, like Thomas said, it gives peace of mind to the user.
> 
> I'm also thinking that if, in the future, we want to supply pre-built rubrics, we might be able to put them here. There might be a better place for it, but it was something I thought of as I was playing around with this. In a pre-built case, you'd want to show the user a preview before importing.
> 
> Edit and sidenote: If we want to talk more about design, can we avoid using GitHub comments? Being that you can't reply in a thread, I find conversations trying to happen with GitHub comments are hard to follow.

I like preview a lot, but it seems like a detail that should come later. If we add it now, we have to make it work well. That is time taken from other areas.

Happy to chat more in the Teacher Tool channel if GitHub isn't working well. I'm also updating in the Figma to represent this idea.

File: teachertool/src/services/fileSystemService.ts (hunk 4 above)
Comment (thsparks) [reply teachertool/src/services/fileSystemService.ts]: I'm happy to have it be exported as unnamed-rubric for now if it hasn't been set yet

**Review by srietkerk — APPROVED** [review]
LGTM
